src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesGcm.Unix.cs

```diff
@@ -0,0 +1,202 @@
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ref nonEmptyArray, out ciphertextBytesWritten, ref nonEmptyArray, plaintext.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherFinalEx(_ctxHandle, ref nonEmptyArray, out int bytesWritten))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                // Should be 0 but let's validate anyway
+                ciphertextBytesWritten += bytesWritten;
+            }
+            
+            if (ciphertextBytesWritten != plaintext.Length)
+            {
+                // this shouldn't happen
+                throw new CryptographicException();
+            }
+
+            if (!Interop.Crypto.EvpAesGcmGetTag(_ctxHandle, ref MemoryMarshal.GetReference(tag), tag.Length))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+        }
+
+        public void Decrypt(ReadOnlySpan<byte> nonce, ReadOnlySpan<byte> ciphertext, ReadOnlySpan<byte> tag, Span<byte> plaintext, ReadOnlySpan<byte> associatedData)
+        {
+            AesAEAD.CheckArguments(nonce, plaintext, ciphertext, tag);
+            CheckParameters(nonce.Length, tag.Length);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+
+            if (!Interop.Crypto.EvpCipherSetKeyAndIV(_ctxHandle, ref nullRef, ref MemoryMarshal.GetReference(nonce), 0 /* decrypting */))
+            {
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+            }
+
+            if (associatedData.Length != 0)
+            {
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ref nullRef, out _, ref MemoryMarshal.GetReference(associatedData), associatedData.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+            }
+
+            int plaintextBytesWritten;
+            if (ciphertext.Length != 0)
+            {
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ref MemoryMarshal.GetReference(plaintext), out plaintextBytesWritten, ref MemoryMarshal.GetReference(ciphertext), ciphertext.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesGcmSetTag(_ctxHandle, ref MemoryMarshal.GetReference(tag), tag.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherFinalEx(_ctxHandle, ref MemoryMarshal.GetReference(plaintext.Slice(plaintextBytesWritten)), out int bytesWritten))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
```

Comment:
Shouldn't this have the invalid tag string?

Reply:
yes, thanks for spotting